watchdog: implemented systemd-style watchdog support

eras opened the pull request:
Implements support for systemd watchdog based on checking that
`global_lock` isn't locked for too long. The support is compiled in if the
libsystemd-daemon is enabled during configure (this existed already)
and it is enabled in the service file with ie.:

    [Service]
    Type=notify
    TimeoutStartSec=10m
    WatchdogSec=5m
    ExecStart=/opt/bin/tvheadend --stderr --syslog --nobackup
    User=tvheadend
    Restart=always

(The example has large timeout due to the long time my setup can take to start.)

The watchdog runs in its own thread and it tries to periodically lock the
`global_lock` which seems be a very good candidate for hanging tvheadend.
If there are other locks to monitor they can be easily added as well.

Saentist commented:
Hope this will not bring continuous restart-loop.

CvH commented:
```
...
Restart=always
StartLimitInterval=60
StartLimitBurst=2
```
^^^^ starts it max two times in 120sec, not sure if this is helpful or harming

mpmc commented:
I like the watchdog support, but me personally I don't think tvh should restart at all on a crash/hang, it should stay dead. However those that do want it should have the option.

> The watchdog runs in its own thread and it tries to periodically lock the
global_lock which seems be a very good candidate for hanging tvheadend.
If there are other locks to monitor they can be easily added as well.

This honestly sounds like a band-aid, not a bad one by any means, but I think the proper solution would be to find the cause? Which might not be an easy task :disappointed:

eras commented on Makefile:
```diff
@@ -260,7 +260,9 @@ SRCS-1 = \
 	src/wizard.c \
-	src/memoryinfo.c
+	src/memoryinfo.c \
+	src/watchdog.c
```
I considered this, but then main.c would have ifdefs instead which I think is worse?

eras replied:
As an alternative I could arrange watchdog.h to have #ifdeffed empty inline definitions for watchdog_* functions. This is probably the most preferred option?

perexg replied:
Yes, add empty inline functions to the header file based on ifdefs. See src/dbus.h for an example.

eras commented on src/watchdog.c:
```diff
@@ -0,0 +1,108 @@
+  watchdog_enabled = sd_watchdog_enabled(0, &watchdog_interval_usec) > 0;
+  if (watchdog_enabled) {
+     /* suggested by sd_watchdog_enabled documentation: */
+    watchdog_interval_usec /= 2;
```
There is no default; to enable watchdog one sets the timeout.

eras commented:
@mpmc I certainly agree it's not a solution for bugs - proper solution for tracking these kinds of issues would involve keeping track of who is locking `global_lock`, but this is more complicated and of course what to do varies by bug. But it also allows to detect the frequency of the problem better by keeping logs of it.

For example I have noticed that the system can run a week or two weeks at a time, but then to crash multiple times on weekends between 9-10 am. This is clearly a hint :). In addition, as systemd uses SIGABRT to stop the process I imagine it's possible to collect these core dumps for analysis with systemd-coredump.

Another thing is that sometimes it might be that the situation is irrecoverable, ie. the DVB input device might require a system reboot to resolve the situation. This also is now enabled (at least in principle) by a watchdog mechanism.

azlm8t commented:
I currently restart by having a crontab curl to the web server and restart tvheadend if it fails to get anything after a minute, so your approach will be better for me, though if it generates a core dump every time is there an automatic cleanup?

Maybe I missed it, but perhaps you could add your sample service file changes to the commit (and raise a bug) since otherwise most people will not know the option exists. Judging by other comments, don't enable it by default in the shipped configuration file.

mpmc commented:
> Maybe I missed it, but perhaps you could add your sample service file changes to the commit (and raise a bug) since otherwise most people will not know the option exists. **Judging by other comments, don't enable it by default in the shipped configuration file.**

My initial concerns were the same as @Saentist's but @CvH gave a solution to that, which could be the default. If the majority want an auto-restart it's fine with me! :+1:

eras commented:
Changes:

- Rebased against current master
- Makefile now uses SRCS-${CONFIG_LIBSYSTEMD_DAEMON}, eliminating an #if from watchdog.c
- Minor fixes, no changes in functionality
- An example on enabling the watchdog can now be found from debian/watchdog
- Updated copyright year. Happy New Year :-).

eras commented:
@azlm8t I don't think core files are produced or collected by default.. But perhaps in other systems they can be. In those cases I hope they already have a mechanism for purging old cores :).

eras commented:
Btw, with the timeouts I propose in debian/tvheadend.service there is no possibility of rapid restarts due to watchdog, because the times are so long.

However, rapid restarts are already possible without this watchdog code, because the file says 'Restart=on-failure' and tvheadend might hit an assertion repeatedly on startup. In those cases it seems systemd catches the situation (there are probably sane defaults already in place) and puts the system in failed state. But perhaps the .service-file could be revised to incorporate @CvH 's solution.

perexg commented:
Merged. Thanks.
